### PR TITLE
Improve Windows Ansible roles

### PR DESCRIPTION
**Pull request by ionutbalutoiu**
General improvements for the Windows Ansible roles. The most notable changes are:

* Use the updated `servicewrapper.exe` and pass the parameters via a config file
* Log rotation for all the Windows agents via the Windows service manager failure actions
* Improve PowerShell scripts logic and add error checking where required
* Improve `set_ip_facts.yml` logic. The default interface chosen is the one with the default gateway set up instead of taking the first interface as seen by Ansible. There is a new option `sdn_preferred_nic_name` to force a particular interface to be used for the SDN setup. This fixes issue https://github.com/openvswitch/ovn-kubernetes/issues/492
* Add option to leave firewall enabled and open the required firewall rules. This fixes the issue https://github.com/openvswitch/ovn-kubernetes/issues/500. Thank-you @lanoxx for identifying the firewall rules.

### CODE REVIEW

File: contrib/ovn-kubernetes-cluster.yml [hunk 1]
```diff
@@ -46,10 +46,10 @@
     - name: Ensure /etc/hosts is updated on kube-master
       lineinfile:
         path: /etc/hosts
-        regexp: '{{ansible_hostname}}$'
-        line: '{{host_public_ip}} {{ansible_hostname}}'
+        state: present
+        line: '{{ host_public_ip }} {{ ansible_hostname | lower }}'
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/ovn-kubernetes-cluster.yml]: Ah, I was not sure about default `state: present`. I was under the impression that `regexp` is redundant, but you're right. I'll add it back.

File: contrib/ovn-kubernetes-cluster.yml [hunk 2]
```diff
@@ -84,16 +86,14 @@
...
         name: windows/kubernetes
     - name: Ensure /etc/hosts is updated on kube-master
       become: true
       become_method: sudo
       lineinfile:
         path: /etc/hosts
-        regexp: '{{ansible_hostname}}$'
-        line: '{{host_public_ip}} {{ansible_hostname}}'
+        line: '{{ host_public_ip }} {{ ansible_hostname | lower }}'
+        state: present
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/ovn-kubernetes-cluster.yml]: ditto

File: contrib/roles/windows/kubernetes/tasks/main.yml [hunk 3]
```diff
@@ -1,64 +1,59 @@
...
-  include_tasks: set_ip_facts.yml
+- name: Kubernetes | Configure the Windows firewall
+  block:
+    - name: Disable firewall (only for testing environments)
+      win_command: NetSh Advfirewall set allprofiles state off
+      when: firewall.disable_completely == True
+
+    - name: Enable firewall (for production environments)
+      win_command: NetSh Advfirewall set allprofiles state on
+      when: firewall.disable_completely == False
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/tasks/main.yml]: Will do

File: contrib/roles/windows/kubernetes/tasks/main.yml [hunk 4]
```diff
@@ -1,64 +1,59 @@
...
-
-- name: Kubernetes | set hns_exists
-  set_fact:
-    hns_exists: "{{ false if hns_net_id_out.stdout == '' else true}}"
-
-- name: Kubernetes | setup sdn
+# This will create the HNS network if it doesn't exist, or get its HNS id if
+# it was already previously created.
+- name: Kubernetes | Setup SDN
   include_tasks: ./setup_sdn.yml
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/tasks/main.yml]: Will do

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_hns.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      }
+      ovs-vsctl.exe --timeout {{ ovs_cmd_timeout }} --no-wait --may-exist add-br "$bridgeName"
+      if($LASTEXITCODE) {
+          Throw "Failed to add the OVS bridge"
+      }
+      ovs-vsctl.exe --timeout {{ ovs_cmd_timeout }} --no-wait --may-exist add-port "$bridgeName" "$($net.NetworkAdapterName)"
+      if($LASTEXITCODE) {
+          Throw "Failed to add the HNS interface to OVS bridge"
+      }
+      Enable-OVSOnHNSNetwork $net.ID
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/tasks/setup_ovs_hns.yml]: Not really. Error checking for `$LASTEXITCODE` is needed when running binaries from the disk.

`Enable-OVSOnHNSNetwork` is a PowerShell function from the `OVS` module. The function itself should handle error checking and throw exception if something is wrong.

File: contrib/roles/windows/kubernetes/tasks/setup_ovs_hns.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      Set-Service "ovs-vswitchd" -StartupType Automatic
+      Start-Service "ovs-vswitchd"
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/tasks/setup_ovs_hns.yml]: `Start-Service` will throw an exception if something is wrong.

We have `$ErrorActionPreference = "Stop"` at the beginning of the script. This means that the script will stop its execution and finish with exit code != 0 if any exception is thrown (and not treated in a try / catch block).

File: contrib/roles/windows/kubernetes/vars/windows.yml [hunk 7]
```diff
@@ -15,6 +17,16 @@ kubernetes_info:
   K8S_API_SERVICE_IP: "{{K8S_API_SERVICE_IP | default('10.0.9.1')}}"
   infracontainername: "ovnkubernetes/pause"
 
-host_info:
-  username: "{{ ansible_user }}"
-  password: "{{ ansible_password }}"
+firewall:
+  disable_completely: true  # WARNING: Do NOT enable this in production. Used only for testing purposes.
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/vars/windows.yml]: I'm working on a commit to have deployments working with firewall enabled by default.

Into another conversation, we agreed to leave it here, if everything works as expected with firewall enabled by default.

File: contrib/roles/windows/kubernetes/tasks/main.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply contrib/roles/windows/kubernetes/tasks/main.yml]: I'll make sure that the default firewall rules is exactly what's needed to have the firewall enabled by default.

**Comment by aserdean** [comment]
Thanks for all the work @ionutbalutoiu . I did a sanity test and everything was ok.

**Pull request merged**